src/mscorlib/src/System/Text/StringBuilder.CoreCLR.cs

```diff
@@ -8,11 +8,91 @@ namespace System.Text
 {
     public partial class StringBuilder
     {
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal extern unsafe void ReplaceBufferInternal(char* newBuffer, int newLength);
+        private int GetReplaceBufferCapacity(int requiredCapacity)
+        {
+            // This function assumes that required capacity will be less 
+            // than the max capacity of the StringBuilder
+            Diagnostics.Debug.Assert(requiredCapacity <= m_MaxCapacity);
+
+            int currentCapacity = Capacity;
+            // round the current capacity to the nearest multiple of 2
+            // that is larger than the required capacity
+            if (currentCapacity < requiredCapacity)
+            {
+                currentCapacity = (requiredCapacity + 1) & ~1;
+            }
+            return currentCapacity;
+        }
+
+        internal unsafe void ReplaceBufferInternal(char* newBuffer, int newLength)
+        {
+            if (newLength > m_MaxCapacity)
+                throw new ArgumentOutOfRangeException("capacity", SR.ArgumentOutOfRange_Capacity);
 
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal extern unsafe void ReplaceBufferAnsiInternal(sbyte* newBuffer, int newLength);
+            m_ChunkChars = new char[GetReplaceBufferCapacity(newLength + 1)];
+
+            fixed (char* buffer = m_ChunkChars)
+            {
+                Buffer.MemoryCopy(newBuffer, buffer, m_ChunkChars.Length * sizeof(char), newLength * sizeof(char));
+                buffer[newLength] = '\0';
+            }
+
+            m_ChunkLength = newLength;
+            m_ChunkPrevious = null;
+            m_ChunkOffset = 0;
+        }
+
+        internal unsafe void ReplaceBufferAnsiInternal(sbyte* newBuffer, int newLength)
+        {
+            if (newLength > m_MaxCapacity)
+                throw new ArgumentOutOfRangeException("capacity", SR.ArgumentOutOfRange_Capacity);
+
+            m_ChunkChars = new char[GetReplaceBufferCapacity(newLength + 1)];
+
+            fixed (char* buffer = m_ChunkChars)
+            {
+#if PLATFORM_UNIX
+                try
+                {
+                    newLength = new UTF8Encoding(true, true).GetChars((byte*)newBuffer, newLength + 1, buffer, m_ChunkChars.Length);
```

Comment:
Suggestion: Could you add /*encoderShouldEmitUTF8Identifier*/ before true? also /*throwOnInvalidBytes*/ before true?

Reply:
I need to change this to use `Encoding.UTF8`. Initially I thought that the original code is trying to produce an empty string when the native buffer contains invalid characters, the only way to achieve that using `UTF8Encoding` is to make it throw exceptions so you need to create your own encoding object.

But as mentioned in my previous post, the original code did not pass `MB_ERR_INVALID_CHARS` so invalid characters were silently converted resulting in a non empty string.